language/vm/src/abi.rs

```diff
@@ -0,0 +1,302 @@
+                    // in module `Child` will continue to run without error. But values of type
+                    // `Parent::T` in `Child` are now "orphaned" in the sense that `Parent` no
+                    // longer exposes any API for reading/writing them.
+                }
+            }
+        }
+
+        // old module's public functions are a subset of the new module's public functions
+        for function in &self.public_functions {
+            if !new_module.public_functions.contains(&function) {
+                struct_and_function_linking = false;
+            }
+        }
+
+        Compatibility {
+            struct_and_function_linking,
+            struct_layout,
+        }
+    }
+
+    /// Return true if `new_module` can safely update `old_module`
+    pub fn can_update(
+        old_module: &ModuleABI,
+        new_module: &CompiledModule,
+        _new_module_dependencies: &[CompiledModule],
+    ) -> bool {
+        // (1) Verify new_module (TODO)
+        // (2) Link new_module against new_module dependencies. (TODO)
+        //     Note: this will *NOT* prevent cylic deps. We need to think about a different scheme
+        //     if we care about this (wich we almost certainly do). One (probably too restrictive)
+        //     solution would be: insist that deps(new_module) are a subset of deps(old_module).
+        //     That would not only prevent cyclic deps, but also preclude the need for linking
+        //     entirely.
+        // (3) Extract the ABI for new_module and check compatibility with old_module
+        old_module
+            .check_compatibility(&ModuleABI::new(new_module))
+            .is_fully_compatible()
+            && panic!("TODO: implement verification, linking, cyclic deps checks")
+    }
+
+    /// Extract an ABI from a `CompiledModule`
+    pub fn new(m: &CompiledModule) -> Self {
+        let id = m.self_id();
+        let structs = m
+            .struct_defs()
+            .iter()
+            .map(|d| StructABI::new(m, d))
+            .collect();
+        let public_functions = m
+            .function_defs()
+            .iter()
+            .filter_map(|f| {
+                if f.is_public {
+                    Some(FunctionABI::new(m, m.function_handle_at(f.function)))
+                } else {
+                    None
+                }
+            })
+            .collect();
+        Self {
+            id,
+            structs,
+            public_functions,
+        }
+    }
+}
+
+impl TypeABI {
+    /// Create a `TypeABI` for `SignatureToken` `s` in module `m`.
+    pub fn new(m: &CompiledModule, s: &SignatureToken) -> Self {
+        use SignatureToken::*;
+        match s {
+            Struct(shi) => {
+                let handle = m.struct_handle_at(*shi);
+                assert!(handle.type_parameters.is_empty(), "A struct with N type parameters should be encoded as StructModuleInstantiation with type_arguments = [TypeParameter(1), ..., TypeParameter(N)]");
```

Comment:
@dariorussi / @vgao1996: double-checking that this is true

Reply:
yes it's true (if that is what you mean) that the handle on a `Struct` has to be a non generic type and a handle on a `StructInstantiation` must be a generic type